### PR TITLE
ECPINT 1983

### PR DESCRIPTION
**Pull request by ved-beejadhur-cko**
**Jira:** [ECPINT-1983](https://checkout.atlassian.net/browse/ECPINT-1983)

### Notes

This Pull Request handles store logging to cloudEvent and on the store

### Housekeeping

- [ ] Unit & Integrations tests
- [ ] Updated relevant documentation
- [ ] Manually tested new functionality

### CODE REVIEW

File: src/Helper/CkoLogger.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+namespace Checkoutcom\Helper;
+
+use Checkoutcom\Config\Config;
+use GuzzleHttp\Psr7\Request;
+use GuzzleHttp\Client;
+use RuntimeException;
+use Checkoutcom\Helper\Utilities;
+use Psr\Log\LoggerInterface;
+
+class CkoLogger {
```
Comment (mitja-kramberger-cko) [inline src/Helper/CkoLogger.php:9]: We can just use `monolog` to support different transports (in monolog language this is a `handler`). Transport being shopware6 logger and our logging service (cloudevent API). We should be able to compose them together.

Documentation:
https://github.com/Seldaek/monolog/blob/main/doc/02-handlers-formatters-processors.md#handlers

Example handler:
- https://github.com/Seldaek/monolog/blob/main/src/Monolog/Handler/LogglyHandler.php
- https://github.com/Seldaek/monolog/blob/main/src/Monolog/Handler/ErrorLogHandler.php

Looks like we only need to implement `write` method which is specific for a handler (transport).

File: src/Helper/CkoLogger.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        $logBody['data']['scope'] = $data['scope'];
+        $logBody['data']['message'] = $data['message'];
+        $logBody['cko']['correlationId'] = $data['id'];
+        $logBody['cko']['loglevel'] = 'info';
+    
+        $header =  [
+            'Content-Type' => 'application/cloudevents+json',
+        ];
+    
+        $url = Url::getCloudEventUrl(config::publicKey());
```
Comment (mitja-kramberger-cko) [inline src/Helper/CkoLogger.php]: We don't need to branch based on the current environment, we can just alwayys post to production API (for cloudevents) and split the traffic based on `source` attribute. Requirement is to be able to zoom into a specific Merchant (store instance).

Screenshot 2021-01-06 at 08 45 37

File: src/Helper/CkoLogger.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            $header,
+            json_encode($logBody)
+        );
+    
+    }
+    
+    /**
+     *  function to get message of the event
+     */
+    public function log($log) {
```
Comment (mitja-kramberger-cko) [inline src/Helper/CkoLogger.php]:
```public function log($log) {
    self::$logger->error(json_encode($log));

    if (config::logcloudEvent()) {
        try {
            self::postLoggingRequest($log);
        } catch (Exception $e) {
            self::$logger->error('Logging API not available');
            self::$logger->error(json_encode($e));
        }
    }
}```

File: src/Helper/Url.php [hunk 4]
```diff
@@ -86,6 +88,16 @@ public function getRetrieveInstrumentUrl(string $customerId) {
         return self::checkUrlSlash(config::cloudPluginUrl()).'customer/' . $customerId;
     }
 
+    /**
+     * cloudEvent logging url
+     */
+    public function getCloudEventUrl(String $publicKey) {
+        $isLive = self::isLive($publicKey);
+        $url = $isLive ? self::CLOUDEVENT_LIVE_URL : self::CLOUDEVENT_QA_URL;
```
Comment (mitja-kramberger-cko) [inline src/Helper/Url.php]: Not needed, we can just point to `CLOUDEVENT_LIVE_URL`

File: src/Helper/Utilities.php [hunk 5]
```diff
@@ -148,4 +146,21 @@ public function fixAmount($amount, $currency = '', $reverse = false)
 
         return $price;
     }
+
+    /**
+     * build log body
+     */
+    public function contructLogBody($error, $scope, $type, $id = false) {
```
Comment (mitja-kramberger-cko) [inline src/Helper/Utilities.php]: Lets utilize Logger levels:

```
$logger->info(...)
$logger->error(...)
$logger->warning(...)
```

https://github.com/Seldaek/monolog/blob/main/doc/01-usage.md#log-levels


Example Logger API:

```
$logger->error([
    "id" => $id,
    "scope" => $scope,
    "message" => json_encode($logBody),
    "type" => $type
])
```

File: src/Storefront/Controller/ComponentsController.php [hunk 6]
```diff
@@ -242,13 +243,19 @@ public function removeCard(SalesChannelContext $context, $card) : Response
...
-            ), 200, ['Content-Type' => 'text/javascript']
-        );
+            return new Response(
+                json_encode(
+                    []
+                ), 200, ['Content-Type' => 'text/javascript']
+            );
+            
+        } catch (\Exception $e) {
+            $logMessage = Utilities::contructLogBody($e, "cko delete instrument", "checkout.intrument.delete.error", $uuid);
```
Comment (mitja-kramberger-cko) [inline src/Storefront/Controller/ComponentsController.php]: The question can we recover (this has to be considered for every use-case)?
If not we should throw an exception (and also log it).

File: src/Handler/DatadogHandler.php [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+                
+                throw new ckoException($e->getMessage(), "Datadog handler log", "Datadog.log.error", "Error");
+            }
+        }
+
+    }
+
+    public function logLevelName($logLevel) {
+
+        $level;
```
Comment (mitja-kramberger-cko) [inline src/Handler/DatadogHandler.php]: We can define a hashmap with mapping, and perform a simple lookup.

```
$errorMapping = array();
$errorMapping[100] = 'debug';
$errorMapping[200] = 'info';

// ...

$errorMapping[$logLevel];
```

File: src/Handler/DatadogHandler.php [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+use GuzzleHttp\Exception\GuzzleException;
+use Koriym\HttpConstants\Method;
+use Koriym\HttpConstants\RequestHeader;
+use Monolog\Handler\AbstractProcessingHandler;
+use Checkoutcom\Helper\Utilities;
+use Checkoutcom\Helper\ckoException;
+use Checkoutcom\Config\Config;
+use Checkoutcom\Helper\Url;
+
+class DatadogHandler extends AbstractProcessingHandler {
```
Comment (mitja-kramberger-cko) [inline src/Handler/DatadogHandler.php]: It should be names `CloudEventsHandler` as we are coding against Datadog.

File: src/Helper/CkoLogger.php [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    public function log($message, $scope, $type, $id = false, $logLevel) {
+
+        $body = Utilities::contructLogBody($message, $scope, $type, $id, $logLevel);
+        
+        $url = Url::getCloudEventUrl();
+        $formatter = new JsonFormatter();
+        
+        // create instance of the datadog handler to log on cloudEvent
+        $datadogLogs = new DatadogHandler($url, $logLevel, true);
+        $datadogLogs->setFormatter($formatter);
```
Comment (mitja-kramberger-cko) [inline src/Helper/CkoLogger.php]: Should the formatter be enabled only in `dev mode`?

File: src/Helper/CkoLogger.php [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+     */
+    public function log($message, $scope, $type, $id = false, $logLevel) {
+
+        $body = Utilities::contructLogBody($message, $scope, $type, $id, $logLevel);
+        
+        $url = Url::getCloudEventUrl();
+        $formatter = new JsonFormatter();
+        
+        // create instance of the datadog handler to log on cloudEvent
+        $datadogLogs = new DatadogHandler($url, $logLevel, true);
```
Comment (mitja-kramberger-cko) [inline src/Helper/CkoLogger.php]: We can optionally enable custom handler, for example:

```
 if (config::logcloudEvent() == true) {
  self::$logger->pushHandler(new DatadogHandler($url, $logLevel, true));
}
```

File: src/Handler/DatadogHandler.php [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        $logBody['data']['scope'] = $obj->scope;
+        $logBody['data']['message'] = $obj->message;
+        $logBody['cko']['correlationId'] =$obj->id;
+        $logBody['cko']['loglevel'] = $logLevel;
+
+        $header =  [
+            'Content-Type' => 'application/cloudevents+json',
+        ];
+
+        if (config::logcloudEvent() == true) {
```
Comment (mitja-kramberger-cko) [inline src/Handler/DatadogHandler.php]: We don't need to check, please look at below comment - for CKOLogger.

File: src/Helper/CkoLogger.php [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+ * 
+ */
+class CkoLogger {
+
+    /**
+     * @var LoggerInterface
+     */
+    public static $logger;
+
+    function __construct(LoggerInterface $logger) {
```
Comment (mitja-kramberger-cko) [inline src/Helper/CkoLogger.php]: In the constructor we need to initialize logger (which needs to be done once)
and then in log, we only proxy to logger itself.

File: src/Helper/CkoLogger.php [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+        self::$logger = $logger;
+    }
+    
+    /**
+     *  log to cloudEvent and shopware 6
+     */
+    public function log($message, $scope, $type, $id = false, $logLevel) {
```
Comment (mitja-kramberger-cko) [inline src/Helper/CkoLogger.php]: `$logLevel` should be an ENUM, you can only provide predefined values (e.g. info, error).

File: src/Helper/CkoLogger.php [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+        $body = Utilities::contructLogBody($message, $scope, $type, $id, $logLevel);
```
Comment (mitja-kramberger-cko) [inline src/Helper/CkoLogger.php]: This can be moved to the Handler.

File: src/Service/PaymentService.php [hunk 15]
```diff
@@ -69,14 +71,13 @@ public function create($param, $correlationId)
             return $response;
             
         } catch (Exception $e) {
-            // @todo catch and log error
-            // return http status code
-            // throw new RuntimeException('cko Payment error : ' . $e->getMessage());
+
+            CkoLogger::log($e->getMessage(), "cko create payment", "checkout.create.payment.error", $correlationId, "Error");
```
Comment (mitja-kramberger-cko) [inline src/Service/PaymentService.php]: CorrelationID should be unique for a single request, so we can set it once (for example in the controller) as:

```CkoLogger::setCorrelationId('XXX');```

Which means we don't need to pass it for every log statement.
Please look at the cloud plugin for similar pattern.

File: src/Service/PaymentService.php [hunk 16]
```diff
@@ -117,25 +118,23 @@ public function checkPayment($id)
...
-                $response['message'] = "An error has occured"; // @todo give proper error message to frontstore
+                $response['message'] = "An error has occured";
             }
 
             return $response;
             
         } catch (Exception $e) {
-            // @todo catch and log error
 
+            CkoLogger::log($e->getMessage(), "cko verify payment", "checkout.payment.verify.error", $id, "Error");
```
Comment (mitja-kramberger-cko) [inline src/Service/PaymentService.php]: Can we directly reference the logger instance, so we can bypass the need to map to a loglevel, and just direct log a message, for example:

```
CkoLogger::loggger->Error(array(
    "type": => "checkout.create.payment.error"
    "scope" => "cko create payment",
    "message" => $e->getMessage()
))
```

File: src/Helper/CkoLogger.php (hunk 12 above)
Comment (mitja-kramberger-cko) [reply src/Helper/CkoLogger.php]: For example:

```
class CkoLogger {

    public static $logger;

    function __construct(LoggerInterface $logger) {
        self::$logger = $logger;

        if (config::logcloudEvent() == true) {
            self::$logger->pushHandler(new DatadogHandler(
                Url::getCloudEventUrl(), $logLevel, true
            ));
        }
    }

    public static function log($message, $scope, $type, $id = false, $logLevel) {
        self::$logger->$logLevel($body);
    }
}
```

**Comment by mitja-kramberger-cko** [comment]
## Functionality for the logger:

- send events to cloudevent API
- enable JSON formatter
- save local file


## This predefined fields are used by CloudEventsHandler and for the logging statements:

```
class LogFields extends Enum
{
    const ID = 'id';
    const MESSAGE = 'message';
    const SCOPE = 'scope';
    const DATA = 'data';
}
```

## Example usage:

```
try {
    return Utilities::postRequest('GET', $url, $header, false);
}
catch (\Exception $e) {
    CkoLogger::logger()->Error([
        LogFields::MESSAGE => $e->getMessage(),
        LogFields::SCOPE => "checkout.payment.instrument",
        LogFields::DATA => [ "customerId" => $exception::$exceptionId ]
    ]);

    throw new RuntimeException($exception::$exceptionMessage)
}
```

## Manual testing

Follow the steps on confluence page:
https://checkout.atlassian.net/wiki/spaces/FS/pages/1845168677/ECP+MW+-+Cloudevents+API+-+Datadog

## TODO

- simplify CloudEventsHandler by using log levels from monolog
- subscribe to all unhandled exception
- log explicitly at point of error

File: src/Helper/Url.php [hunk 17]
```diff
@@ -13,6 +13,7 @@ class Url {
      * url
      */
     public const CKO_IFRAME_URL = "https://cdn.checkout.com/js/framesv2.min.js";
+    public const CLOUDEVENT_LIVE_URL = "https://cloudevents.cko-qa.ckotech.co/logging";
```
Comment (cko-moho-khaleqi) [inline src/Helper/Url.php]: Prod URL is `https://cloudevents.integration.checkout.com`

Comment (avish-bisbeehurry-cko) [reply src/Helper/Url.php]: Url updated to prod

**Review by cko-moho-khaleqi — APPROVED** [review]

**Pull request merged**